### PR TITLE
Store Status of a switch in a separate table

### LINKED ISSUE #263: Save switch status responses in separate table
This events happen very often and they should not interfere with regular switch write operations, which are way more important to proceed.

### PR DESCRIPTION
**Pull request by majst01**
closes #263 

This changes the return type from `SwitchResponse` to `SwitchNotifyResponse`. The caller (metal-core) does not do anything with the response object anyway. The metal-core still works without modifications. The sole purpose was to be able to have a rest unit test with a testable interface.

The state actually stored in the Switch object is deleted and not migrated to the new table.

Performance of the `/v1/metal/switch/notify` request drops from ~35ms -> ~2.5ms in our test environment. This is because we neither fetch, nor store the whole switch object anymore. CPU of metal-api also dropped as a side effect.

image

### CODE REVIEW

File: cmd/metal-api/internal/datastore/switch.go [hunk 1]
```diff
@@ -196,3 +196,39 @@ func (rs *RethinkStore) ConnectMachineWithSwitches(m *metal.Machine) error {
 
 	return nil
 }
+
+// GetSwitchStatus get SwitchStatus for a given switch id
+func (rs *RethinkStore) GetSwitchStatus(id string) (*metal.SwitchStatus, error) {
+	rs.log.Infow("get state", "id", id)
```
Comment (Gerrit91) [inline cmd/metal-api/internal/datastore/switch.go:202]: These kinds of logs should be debug only. metal-api primarily logs requests only.

File: cmd/metal-api/internal/datastore/switch.go [hunk 2]
```diff
@@ -196,3 +196,39 @@ func (rs *RethinkStore) ConnectMachineWithSwitches(m *metal.Machine) error {
...
+		createErr := rs.createEntity(rs.switchStatusTable(), state)
+		if err != nil {
+			return nil, createErr
+		}
+		return state, nil
+	}
+	rs.log.Infow("update state", "old", ss, "new", state)
+	err = rs.updateEntity(rs.switchStatusTable(), state, &ss)
+	return state, err
+}
```
Comment (Gerrit91) [inline cmd/metal-api/internal/datastore/switch.go:234]: You can just use upsertEntity (see event insertions). Should reduce this function to a single line.

File: cmd/metal-api/internal/datastore/switch.go [hunk 3]
```diff
@@ -196,3 +196,39 @@ func (rs *RethinkStore) ConnectMachineWithSwitches(m *metal.Machine) error {
...
+func (rs *RethinkStore) GetSwitchStatus(id string) (*metal.SwitchStatus, error) {
+	rs.log.Infow("get state", "id", id)
+	var ss metal.SwitchStatus
+	err := rs.findEntityByID(rs.switchStatusTable(), &ss, id)
+	if err != nil && !metal.IsNotFound(err) {
+		return nil, err
+	}
+	if err != nil && metal.IsNotFound(err) {
+		return &metal.SwitchStatus{}, nil
+	}
```
Comment (Gerrit91) [inline cmd/metal-api/internal/datastore/switch.go:210]: For consistency, I think this kind of error handling should go into the service layer. Other functions of this package, which do similar things, do not return empty responses on non-existent entities.

File: cmd/metal-api/internal/service/v1/switch.go [hunk 4]
```diff
@@ -92,26 +98,30 @@ type SwitchSync struct {
...
 
-func NewSwitchResponse(s *metal.Switch, p *metal.Partition, nics SwitchNics, cons []SwitchConnection) *SwitchResponse {
+func NewSwitchResponse(s *metal.Switch, ss *metal.SwitchStatus, p *metal.Partition, nics SwitchNics, cons []SwitchConnection) *SwitchResponse {
 	if s == nil {
 		return nil
 	}
 
 	var lastSync *SwitchSync
-	if s.LastSync != nil {
+	if ss.LastSync != nil {
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/v1/switch.go:107]: `ss == nil` should also be checked.

File: cmd/metal-api/internal/service/switch-service.go [hunk 5]
```diff
@@ -210,37 +210,35 @@ func (r *switchResource) notifySwitch(request *restful.Request, response *restfu
...
 		r.sendError(request, response, defaultError(err))
 		return
 	}
 
-	resp, err := makeSwitchResponse(s, r.ds)
-	if err != nil {
-		r.sendError(request, response, defaultError(err))
-		return
-	}
+	r.log.Infow("switchservice notify", "state", newSS, "resp", resp)
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/switch-service.go:241]: Should be debug level.

File: cmd/metal-api/internal/service/switch-service.go [hunk 6]
```diff
@@ -628,8 +626,12 @@ func makeSwitchResponse(s *metal.Switch, ds *datastore.RethinkStore) (*v1.Switch
 
 	nics := makeSwitchNics(s, ips, machines)
 	cons := makeSwitchCons(s)
+	ss, err := ds.GetSwitchStatus(s.ID)
+	if err != nil {
+		return nil, err
+	}
```
Comment (Gerrit91) [inline cmd/metal-api/internal/service/switch-service.go:632]: Should go into `findSwitchReferencedEntites`, which is where the other references entities are gathered.

**Review by Gerrit91 — COMMENTED** [review]
Very reasonable addition! Left you some comments.

File: cmd/metal-api/internal/datastore/switch.go (hunk 2 above)
Comment (majst01) [reply cmd/metal-api/internal/datastore/switch.go:234]: But upsert does not create the entity if it does not exists, or do i miss something ?

Comment (Gerrit91) [reply cmd/metal-api/internal/datastore/switch.go:234]: I think it does, it uses the same rethinkdb function as for creating entities, but does a replace on existing documents.

File: cmd/metal-api/internal/datastore/switch.go (hunk 1 above)
Comment (majst01) [reply cmd/metal-api/internal/datastore/switch.go]: done

File: cmd/metal-api/internal/datastore/switch.go (hunk 2 above)
Comment (majst01) [reply cmd/metal-api/internal/datastore/switch.go]: done, and fixed the BUG in upsert not setting creation date

File: cmd/metal-api/internal/datastore/switch.go (hunk 3 above)
Comment (majst01) [reply cmd/metal-api/internal/datastore/switch.go]: done

File: cmd/metal-api/internal/service/v1/switch.go (hunk 4 above)
Comment (majst01) [reply cmd/metal-api/internal/service/v1/switch.go]: done

File: cmd/metal-api/internal/service/switch-service.go (hunk 5 above)
Comment (majst01) [reply cmd/metal-api/internal/service/switch-service.go]: removed

File: cmd/metal-api/internal/service/switch-service.go (hunk 6 above)
Comment (majst01) [reply cmd/metal-api/internal/service/switch-service.go]: done

**Review by Gerrit91 — APPROVED** [review]

**Review by mwennrich — APPROVED** [review]

**Pull request merged**